File: MyFirstApp/app/src/main/java/com/example/myfirstapp/MainActivity.java
package com.example.myfirstapp;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbManager;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.felhr.usbserial.UsbSerialDevice;
import com.felhr.usbserial.UsbSerialInterface;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Map;

public class MainActivity extends AppCompatActivity {
    public static final String EXTRA_MESSAGE = "com.example.myfirstapp.MESSAGE";
    public static final String ACTION_USB_PERMISSION = "com.example.myfirstapp.USB_PERMISSION";
    public static final String ACTION_USB_DEVICE_ATTACHED = "com.example.myfirstapp.ACTION_USB_DEVICE_ATTACHED";

    private UsbManager usbManager;
    private UsbDevice usbDevice;
    private UsbDeviceConnection usbConnection;
    private UsbSerialDevice usbSerialDevice;
    private final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            UsbDevice usbDevice = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE);
            usbConnection = usbManager.openDevice(usbDevice);
            usbSerialDevice = UsbSerialDevice.createUsbSerialDevice(usbDevice, usbConnection);
            if (usbSerialDevice != null) {
                if (usbSerialDevice.open()) { //Set Serial Connection Parameters.
                    usbSerialDevice.setBaudRate(9600);
                    usbSerialDevice.setDataBits(UsbSerialInterface.DATA_BITS_8);
                    usbSerialDevice.setStopBits(UsbSerialInterface.STOP_BITS_1);
                    usbSerialDevice.setParity(UsbSerialInterface.PARITY_NONE);
                    usbSerialDevice.setFlowControl(UsbSerialInterface.FLOW_CONTROL_OFF);
                    usbSerialDevice.read(mCallback);
                    updateTextView("Serial Connection Opened!\n");
                }
            }
        }

        UsbSerialInterface.UsbReadCallback mCallback = new UsbSerialInterface.UsbReadCallback() {
            //Defining a Callback which triggers whenever data is read.
            @Override
            public void onReceivedData(byte[] arg0) {
                String data = null;
                try {
                    data = new String(arg0, "UTF-8");
                    data.concat("/n");
                    updateTextView(data);
                } catch (UnsupportedEncodingException e) {
                    // Oops
                }
            }
        };
    };
    private boolean connected = false;

    public void updateTextView(String data) {
        TextView dataStream = findViewById(R.id.dataStream);
        String existingData = dataStream.getText().toString();
        existingData.concat("/n");
        existingData.concat(data);
        dataStream.setText(existingData);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        usbManager = (UsbManager) getSystemService(Context.USB_SERVICE);
        IntentFilter filter = new IntentFilter(ACTION_USB_DEVICE_ATTACHED);
        registerReceiver(broadcastReceiver, filter);
    }

    private void connectUsbDevice() {
        HashMap<String, UsbDevice> usbDevices = usbManager.getDeviceList();
        for (Map.Entry<String, UsbDevice>entry: usbDevices.entrySet()) {
            usbDevice = entry.getValue();
            int vendorId = usbDevice.getVendorId();
            if (vendorId != 0) {
                PendingIntent pi = PendingIntent.getBroadcast(this, 0, new Intent(ACTION_USB_PERMISSION), 0);
                usbManager.requestPermission(usbDevice, pi);
                Button connectButton = findViewById(R.id.connectButton);
                connectButton.setText(getResources().getString(R.string.button_disconnect));
                connected = true;
            }
        }
    }

    private void disconnectUsbDevice() {
        if (usbSerialDevice != null) {
            usbSerialDevice.close();
        }
        Button connectButton = findViewById(R.id.connectButton);
        connectButton.setText(getResources().getString(R.string.button_connect));
        connected = false;
    }

    /** Called when the user toggles the USB connection on or off */
    public void onToggleConnect(View view) {
        if (!connected) {
            // Perform actual USB connection if there are devices connected
            HashMap<String, UsbDevice> usbDevices = usbManager.getDeviceList();
            if (!usbDevices.isEmpty()) {
                connectUsbDevice();
            } else {
                new AlertDialog.Builder(this)
                        .setTitle(getResources().getString(R.string.dialog_no_connection_title))
                        .setMessage(getResources().getString(R.string.dialog_no_connection_message))
                        .setNeutralButton(getResources().getString(R.string.button_ok), null)
                        .show();
            }
        } else {
            new AlertDialog.Builder(this)
                    .setTitle(getResources().getString(R.string.dialog_disconnect_title))
                    .setMessage(getResources().getString(R.string.dialog_disconnect_message))
                    .setPositiveButton(getResources().getString(R.string.button_yes), new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            disconnectUsbDevice();
                        }
                    })
                    .setNegativeButton(getResources().getString(R.string.button_no), null)
                    .show();
        }
    }
}

File: MyApplication3/app/src/main/java/com/example/myapplication3/MainActivity.kt
package com.example.myapplication3

import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbDeviceConnection
import android.hardware.usb.UsbManager
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentActivity
import com.felhr.usbserial.UsbSerialDevice
import com.felhr.usbserial.UsbSerialInterface
import com.felhr.usbserial.UsbSerialInterface.UsbReadCallback
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity() {
    lateinit var m_usbManager:UsbManager
    var m_device: UsbDevice? = null
    var m_serial: UsbSerialDevice? = null
    var m_connection: UsbDeviceConnection? = null

    val ACTION_USB_PERMISSION = "permission"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        m_usbManager = getSystemService(Context.USB_SERVICE) as UsbManager
        val filter = IntentFilter()
        filter.addAction(ACTION_USB_PERMISSION)
        filter.addAction(UsbManager.ACTION_USB_ACCESSORY_DETACHED)
        filter.addAction(UsbManager.ACTION_USB_DEVICE_ATTACHED)
        registerReceiver(broadcastReceiver, filter)

/*        val mCallback = UsbReadCallback { data: ByteArray? ->
            val dataStr = String(data!!, charset("UTF-8"))
            Log.i("serial", "Data received: $dataStr")
        }
        */
        //        on.setOnClickListener{sendData("o")}
//        off.setOnClickListener{sendData("x")}
        //sendData("0x38")


        //serialPort.syncWrite(0x38,0)
        // m_serial?.syncWrite("0x38".toByteArray(), 60)
        on.setOnClickListener{sendData("o")}
        off.setOnClickListener{sendData("x")}
        disconnect.setOnClickListener{disconnect()}
        connect.setOnClickListener{startUsbConnecting()}
    }

    private fun startUsbConnecting(){
        val usbDevices:HashMap<String,UsbDevice>? = m_usbManager.deviceList
        if (!usbDevices?.isEmpty()!!){
            var keep = true
            usbDevices.forEach { entry ->
                m_device = entry.value
                val deviceVendorId: Int? = m_device?.vendorId
                Log.i("serial", "vendorId: " + deviceVendorId)
                if (deviceVendorId == 9025) { //needs to match arduino vendor id
                    val intent: PendingIntent = PendingIntent.getBroadcast(this, 0, Intent(ACTION_USB_PERMISSION), 0)
                    m_usbManager.requestPermission(m_device, intent)
                    keep = false
                    Log.i("serial", "connection successful")
                } else {
                    m_connection = null
                    m_device = null
                    Log.i("serial", "unable to connect to device, unknown device id: " + deviceVendorId)
                }
                if (!keep) {
                    return
                }
            }
        }else{
            Log.i("serial", "no usb device connected")
        }
    }

    private fun sendData(input: String){
        m_serial?.write(input.toByteArray())
        Log.i("serial","sending data: " + input.toByteArray())
    }

    private fun disconnect(){
        m_serial?.close()
    }

    private val broadcastReceiver = object :BroadcastReceiver(){
        override fun onReceive(context: Context?, intent: Intent?) {
            if (intent?.action!! == ACTION_USB_PERMISSION){
                val granted: Boolean = intent.extras!!.getBoolean((UsbManager.EXTRA_PERMISSION_GRANTED))
                if (granted){
                    m_connection = m_usbManager.openDevice((m_device));
                    m_serial = UsbSerialDevice.createUsbSerialDevice(m_device, m_connection)
                    if (m_serial != null){
                        if(m_serial!!.open()){
                            m_serial!!.setBaudRate(9600)
                            m_serial!!.setDataBits(UsbSerialInterface.DATA_BITS_8)
                            m_serial!!.setStopBits(UsbSerialInterface.STOP_BITS_1)
                            m_serial!!.setParity(UsbSerialInterface.PARITY_NONE)
                            m_serial!!.setFlowControl(UsbSerialInterface.FLOW_CONTROL_OFF)

                        }
                        else {
                            Log.i("Serial", "port not open")
                        }
                    }
                    else {
                        Log.i("Serial", "port is null")
                    }
                }else{
                    Log.i("Serial", "permission not granted")
                }
            }else if(intent.action == UsbManager.ACTION_USB_ACCESSORY_ATTACHED){
                startUsbConnecting()
            }else if(intent.action == UsbManager.ACTION_USB_DEVICE_DETACHED){
                disconnect()
            }
        }
    }
}
